source/web/testing/TestManager.h

```diff
@@ -0,0 +1,274 @@
+    }
+
+    /// Cleanup test runner specified by runner_id.
+    void CleanupTest(size_t runner_id) {
+      emp_assert(runner_id < test_runners.size());
+      test_runners[runner_id].cleanup();
+    }
+
+    /// Cleanup all test runners.
+    void Cleanup() {
+      // finished running all tests. Make sure all dynamically allocated memory is cleaned up.
+      std::for_each(
+        test_runners.begin(),
+        test_runners.end(),
+        [](TestRunner & runner) {
+          emp_assert(runner.done);
+          if (runner.test != nullptr) runner.test.Delete();
+        }
+      );
+      // Clear test runners.
+      test_runners.clear();
+    }
+
+  public:
+
+    TestManager() {
+      next_test_js_func_id = emp::JSWrap(
+        [this]() { this->NextTest(); },
+        "NextTest"
+      );
+      cleanup_test_js_func_id = emp::JSWrap(
+        [this](size_t test_id) { this->CleanupTest(test_id); },
+        "CleanupTest"
+      );
+      cleanup_all_js_func_id = emp::JSWrap(
+        [this]() { this->Cleanup(); },
+        "CleanupManager"
+      );
+    }
+
+    ~TestManager() {
+      Cleanup();
+      emp::JSDelete(next_test_js_func_id);
+      emp::JSDelete(cleanup_test_js_func_id);
+      emp::JSDelete(cleanup_all_js_func_id);
+    }
+
+    /// Add a test type to be run. The TestManager creates, runs, and cleans up each test.
+    /// This function should be called with the test type (which should inherit from BaseTest) as a
+    /// template argument (e.g., AddTest<TEST_TYPE>(...) ).
+    /// Arguments:
+    /// - test_name specifies the name of the test (this is only used when printing which test is running
+    ///   and does not need to be unique across tests).
+    /// - All subsequent arguments are forwarded to the TEST_TYPE constructor.
+    // For variatic capture: https://stackoverflow.com/questions/47496358/c-lambdas-how-to-capture-variadic-parameter-pack-from-the-upper-scope
+    //  - NOTE: this can get cleaned up quite a bit w/C++ 20!
+    template<typename TEST_TYPE, typename... Args>
+    void AddTest(const std::string & test_name, Args&&... constructor_args) {
+      const size_t runner_id = test_runners.size();
+
+      // create a new test runner for this test
+      test_runners.emplace_back();
+      auto & runner = test_runners[runner_id];
+
+      // name it!
+      runner.test_name = test_name;
+
+      // configure the function that, when called, will create a new instance of TEST_TYPE (using forwarded
+      // arguments as constructor arguments) and then call TEST_TYPE::Setup.
+      runner.create = [constructor_args = std::make_tuple(std::forward<Args>(constructor_args)...), runner_id, this]() mutable {
+        std::apply([this, runner_id](auto&&... constructor_args) {
+          // Record the before test error count
+          this->test_runners[runner_id].before_test_error_count = emp::GetUnitTestOutput().errors;
```

Comment:
Is emp::GetUnitTestOutput() returning just a `struct`? Might be nice to have a class ... also, looks like we're only calling GetUnitTestOutput() to check for .errors... maybe make a function that just says whether there are errors (or how many there are if error count is kept in design)

Reply:
again, this only exists to support Charles' testing macros. I'm going to say it's outside the scope of this pull request to clean the custom unit testing (tests2/unit_tests.h) up.

Reply:
ah, I was wondering where those calls were going to